Step method auto-assignment

fonnesbeck opened the pull request:
This is a first pass at having `sample` choose appropriate step methods for unassigned variables, addressing #832. Each step method has a `competence` static method that is used to evaluate its suitability for a given variable. The assignment logic is in a function called from `sample`. It appears to successfully assign step methods, but sampling is not yet working, so there is much debugging to be done.

Commit:
Draft of step method auto-assignment

Commit:
Assigned selected steps to variables

fonnesbeck commented:
OK, step method assignment appears to be working now. I've left a print statement in for the time being, to get feedback on what is going on. We can take it out before merging.

Please try this on various models. I've changed the `disaster_model` example to auto-assign.

One hitch is that if there is already a particular step method with assigned variables, the auto-assignment will instantiate a new one if additional variables use the same step method, which is not ideal. This is due to the fact that there is not currently a clean way of _adding_ variables to an instantiated step method.

Commit:
Fixed CompoundStep bug with auto-assignment

fonnesbeck commented:
We can tweak the competence values for each step method to our liking. At the moment, it recommends `NUTS` if there are continuous variables of size 5 or greater. Assigns `BinaryMetropolis` to boolean types and `Continuous` variables with k=1. Essentially picks `Metropolis` otherwise.

Commit:
Fixed selected_steps dict append bug; adjusted BinaryMetropolis competence

jsalvatier commented on pymc3/sampling.py:
```diff
@@ -10,8 +10,40 @@
 
-
-def sample(draws, step, start=None, trace=None, chain=0, njobs=1, tune=None,
+def assign_step_methods(model, step):
```
Seems like step should be either a single step or a list.

jsalvatier commented:
Love this. 

I'm not super pumped about having a global `step_methods.step_method_registry` that other people are inevitably going to mutate. I would rather we pass this in and have a default. 

I'm actually a fan of the print statement or something like it. I think maybe we want to move to having a fair bit of judiciously chosen informational output by default since often fitting models doesn't work well the first time. This fits into that. 

I'm thinking things like, step methods chosen, minor info about which step is taking place, efficiency metrics from samplers, how good the starting point found by find_MAP is etc.

jsalvatier commented on pymc3/sampling.py:
```diff
@@ -10,8 +10,40 @@
+    if len(steps)==1:
+        steps = steps[0]    
+                
+    return model, steps
```
why return model here? seems like we want to be able to say `sample(300, step=assign_step_methods(model, []))`

fonnesbeck replied:
Ah, you're right. When I started writing it I assumed I would have to modify the model.

jsalvatier commented on pymc3/step_methods/nuts.py:
```diff
@@ -136,6 +137,14 @@ def astep(self, q0):
+    @staticmethod
+    def competence(var):
+        if var.dtype in continuous_types:
+            if sum(var.shape) > 5:
```
One thing to note is that if a model has 10 continuous variables which are independently specified it will end up choosing metropolis not  NUTS. This is one reason why we might want to have the competence function take in _all_ of the variables not consider them one at a time. I'm leaning in the direction that its better to do it this way.

fonnesbeck replied:
Will have to think about how to implement this.

fonnesbeck replied:
Simplest solution is to just prefer NUTS for any continuous variable, and let users pick Metropolis manually if they want it.

jsalvatier commented:
Excited about this! :+1:

Commit:
assign_step_method only returns step list

walterreade commented:
Wouldn't it be nearly as effective to have a "Guild to Choosing Step Methods" on the documentation page? Does this exist anywhere?

jsalvatier commented:
That's probably good to have in addition, but I think its probably good to
have both since it makes building and fitting models easier.

On Sun, Sep 20, 2015 at 4:54 PM Walter Reade notifications@github.com
wrote:

> Wouldn't it be nearly as effective to have a "Guild to Choosing Step
> Methods" on the documentation page? Does this exist anywhere?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/pymc-devs/pymc3/pull/833#issuecomment-141846057.

Commit:
Replaced step_method_registry with default argument tuple

Commit:
Fixed step_registry reference bug

Commit:
Added vars as attribute to ArrayStepShared so that it may be queried by sample()

Commit:
Fixed step assignment failure in sample()

Commit:
Changed NUTS competence to make it preferred for continuous variables

Commit:
Updated disaster_model_missing to partially auto-allocate step methods

fonnesbeck commented:
Is this merge-able? Any reservations about what I have done here? In the meantime, I will write a couple of tests.

twiecki commented on pymc3/examples/disaster_model_missing.py:
```diff
@@ -55,12 +55,10 @@ def run(n=1000):
 
-        # Use slice sampler for means
+        # Use slice sampler for means (other varibles auto-selected)
         step1 = Slice([early_mean, late_mean])
```
maybe rename to just `step`?

fonnesbeck replied:
Done

twiecki commented on pymc3/sampling.py:
```diff
@@ -4,14 +4,48 @@
+                    assigned_vars = assigned_vars | set(m.vars)
+    
+    # Use competence classmethods to select step methods for remaining variables
+    selected_steps = {s:[] for s in methods}
```
Can use `defaultdict(list)` here.

fonnesbeck replied:
Of course, yes. Done.

twiecki commented on pymc3/step_methods/hmc.py:
```diff
@@ -92,7 +93,12 @@ def astep(self, q0, logp, dlogp):
+    @staticmethod
+    def competence(var):
+        if var in discrete_types:
+            return 0
```
Not sure I'm a fan of using ints here. How about using enums?

fonnesbeck replied:
From the  `enum` package? Can they be ordered? We need an ordered set.

twiecki replied:
not necessarily, can just be a few globals:
`DISCRETE = 0` to enhance readability.

Commit:
Fixed bug in BinaryMetropolis competence; added test for assign_step_methods

Commit:
Rename step1 in disaster_model_missing; used defaultdict in assign_step_methods

Commit:
Implemented IntEnum for competence classification

Commit:
Fix for TravisCI failure related to enum on legacy Python

Commit:
Tweak to .travis.yml edits

jsalvatier commented on pymc3/sampling.py:
```diff
@@ -55,6 +92,10 @@ def sample(draws, step, start=None, trace=None, chain=0, njobs=1, tune=None,
     """
+    model = modelcontext(model)
+    
+    step = assign_step_methods(model, step)
```
One worry is that this will always assign step methods. I've sometimes wanted to sample only a few variables and leave the others fixed. This seems like that makes that hard. 

One fix would be to supply a 'do nothing' sampler. There's probably other things we could do too.

fonnesbeck replied:
I typically just set those variables to constants when I want to do that.

fonnesbeck replied:
OK, check out bdc9256. I have added a `Constant` step method that will simply keep its variables at its current value. Its not included in the default set of step methods searched by `assign_step_method`, so it will never be selected automatically.

Commit:
Fixed TravisCI failure in test_step

Commit:
Added docstring to slice sampler

Commit:
Added Constant step method and test

fonnesbeck commented:
With the addition of the `Constant` step method, how does this look? Any outstanding issues?

twiecki commented:
LGTM. This is a great feature.

jsalvatier commented:
I think this is good. I think maybe we should mention somewhere in the docs or something that you can inspect/modify the assigned methods by using the function `assign_step_methods`. 

Also, have we auto generated the docs? Does that work?

fonnesbeck commented:
I can add something to the tutorial. Regarding auto-generation, are you talking about the docs, or the API? I believe @twiecki used `mkdocs` for the former, and we have yet to tackle the latter.

Commit:
Documented auto step method assignment

- added docstring to assign_step_method
- added paragraph and examples of auto-assignment to tutorial
- added missing dataset for tutorial to docs directory

jsalvatier commented:
I suppose I mean the API, the distributions, functions etc. We have a lot of docstrings it would be nice to put them up somewhere so that they're easily searchable.

Commit:
Truncated S&P500 data for getting_started

fonnesbeck commented:
There are a [few different tools for auto-API generation](https://wiki.python.org/moin/DocumentationTools), but its been awhile since I've looked at them.

Commit:
Debugging getting_started notebook

fonnesbeck commented:
Can someone check out the `getting_started` notebook in the docs? I've yet to successfully have it run to completion. It simply hangs, pegging the CPU on 2 different machines I've tried it on. We may want to consider having a verbose debugging mode for `sample`.

fonnesbeck commented:
The notebooks issue seems unrelated to this pull request, so I will merge and create an issue for it.

jsalvatier commented:
:+1: :smile:
